beam/src/main/java/org/apache/iceberg/beam/FileWriter.java

```diff
@@ -0,0 +1,178 @@
+  private transient BoundedWindow lastSeenWindow;
+
+  public FileWriter(
+      TableIdentifier tableIdentifier,
+      Schema schema,
+      PartitionSpec spec,
+      String hiveMetastoreUrl,
+      Map<String, String> properties
+  ) {
+    this.tableIdentifier = tableIdentifier;
+    this.spec = spec;
+    this.hiveMetastoreUrl = hiveMetastoreUrl;
+    this.schema = schema;
+    this.properties = properties;
+  }
+
+  @StartBundle
+  public void startBundle(StartBundleContext sbc) {
+    start();
+  }
+
+  @ProcessElement
+  public void processElement(ProcessContext context, BoundedWindow window) {
+    appendRecord(
+        context.element(),
+        window,
+        (int) context.pane().getIndex(),
+        context.pane().getIndex()
+    );
+  }
+
+  @FinishBundle
+  public void finishBundle(FinishBundleContext fbc) {
+    DataFile[] files = finish();
+    for (DataFile file : files) {
+      fbc.output(file, Instant.now(), lastSeenWindow);
+    }
+  }
+
+  @VisibleForTesting
+  public void start() {
+    Configuration conf = new Configuration();
+    for (String key : this.properties.keySet()) {
+      conf.set(key, this.properties.get(key));
+    }
+    catalog = new HiveCatalog(
+        HiveCatalog.DEFAULT_NAME,
+        this.hiveMetastoreUrl,
+        1,
+        conf
+    );
+    Table table = HiveCatalogHelper.loadOrCreateTable(
+        catalog,
+        tableIdentifier,
+        schema,
+        spec,
+        properties
+    );
+    this.schema = table.schema();
+    this.locations = table.locationProvider();
+    this.properties = table.properties();
+    this.io = table.io();
+    this.encryptionManager = table.encryption();
+    String formatString = table.properties().getOrDefault(
+        TableProperties.DEFAULT_FILE_FORMAT, TableProperties.DEFAULT_FILE_FORMAT_DEFAULT);
+    this.fileFormat = FileFormat.valueOf(formatString.toUpperCase(Locale.ENGLISH));
+  }
+
+  @VisibleForTesting
+  public void appendRecord(T element, BoundedWindow window, int partitionId, long taskId) {
+    if (writer == null) {
+      LOG.info("Setting up the writer");
+      // We would rather do this in the startBundle, but we don't know the pane
+
+      BeamAppenderFactory<T> appenderFactory = new BeamAppenderFactory<>(schema, properties, spec);
+      OutputFileFactory fileFactory = new OutputFileFactory(
+          spec, fileFormat, locations, io, encryptionManager, partitionId, taskId);
+
+      if (spec.isUnpartitioned()) {
+        writer = new UnpartitionedWriter<>(spec, fileFormat, appenderFactory, fileFactory, io, Long.MAX_VALUE);
+      } else {
+        writer = new PartitionedWriter<T>(
+            spec, fileFormat, appenderFactory, fileFactory, io, Long.MAX_VALUE) {
+          @Override
+          protected PartitionKey partition(T row) {
+            return new PartitionKey(spec, schema);
+          }
+        };
+      }
+    }
+    try {
+      lastSeenWindow = window;
+      writer.write(element);
```

Comment:
What if we get records for multiple partitions? I think `PartitionedWriter` will fail with a runtime exception.